File: Lib/Telerik/Examples/QuickStart/Sparkline/FirstLook/SparklineUserControl.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Drawing;
using System.Data;
using System.Linq;
using System.Text;
using System.Windows.Forms;
using Telerik.WinControls.UI;
using Telerik.WinControls;

namespace Telerik.Examples.WinControls.Sparkline.FirstLook
{
    public partial class SparklineUserControl : UserControl
    {
        private CallType callType;
        private Timer timer;
        private SparkCartesianSeries series;
        private CalleeViewModel owner;

        public SparklineUserControl()
        {
            InitializeComponent();

            FontFamily webUIFont = ThemeResolutionService.GetCustomFont("TelerikWebUI");
            this.alertLabel.LabelElement.CustomFont = webUIFont.Name;
            this.alertLabel.LabelElement.CustomFontSize = 8;
            this.alertLabel.ForeColor = Color.Red;
            this.alertLabel.Text = "\ue403";

            this.timeLabel.Font = new Font("Segoe UI", 9, FontStyle.Regular);
            this.timer = new Timer();
            this.timer.Interval = 750;
            this.timer.Tick += this.Timer_Tick;
        }

        [Browsable(false)]
        [DesignerSerializationVisibility(DesignerSerializationVisibility.Hidden)]
        public CalleeViewModel Owner
        {
            get
            {
                return this.owner;
            }
            set
            {
                if (this.owner != value)
                {
                    this.owner = value;
                }
            }
        }

        [Browsable(false)]
        [DesignerSerializationVisibility(DesignerSerializationVisibility.Hidden)]
        public CallType CallType
        {
            get
            {
                return this.callType;
            }
            set
            {
                if (this.callType != value || this.radSparkline1.Series == null)
                {
                    this.timer.Stop();

                    this.callType = value;
                    this.CreateSeries();
                    this.UpdateSeriesStyles();

                    this.timer.Start();
                }
            }
        }

        public override string ToString()
        {
            switch (this.CallType)
            {
                case CallType.Duration:
                    return "Call Duration";
                case CallType.HoldTime:
                    return "Hold Time";
                case CallType.ResolvedIssues:
                    return "Issues Resolved"; ;
                default:
                    return base.ToString();
            }

        }

        protected override void OnLoad(EventArgs e)
        {
            base.OnLoad(e);

            this.timer.Start();
        }

        private void Timer_Tick(object sender, EventArgs e)
        {
            if (this.Owner == null)
            {
                return;
            }

            this.Owner.AddRandomCall();

            if (this.CallType == CallType.Duration)
            {
                this.alertLabel.LabelElement.Visibility = this.Owner.AverageDuration.Minutes >= 20 ? ElementVisibility.Visible : ElementVisibility.Hidden;
                this.timeLabel.Text = this.Owner.AverageDuration.ToString();
                this.timeLabel.ForeColor = this.Owner.AverageDuration.Minutes >= 20 ? Color.Red : Color.Empty;
            }
            else if (this.CallType == CallType.HoldTime)
            {
                this.alertLabel.LabelElement.Visibility = this.Owner.AverageHoldTime.Minutes >= 10 ? ElementVisibility.Visible : ElementVisibility.Hidden;
                this.timeLabel.Text = this.Owner.AverageHoldTime.ToString();
                this.timeLabel.ForeColor = this.Owner.AverageHoldTime.Minutes >= 10 ? Color.Red : Color.Empty;
            }
        }

        private void CreateSeries()
        {
            this.radSparkline1.Series = null;
            this.series = this.GetSeries();
            this.series.ValueMember = this.GetValueMember(this.CallType);
            this.series.DataSource = this.Owner.Calls;

            this.radSparkline1.Series = series;
        }

        private SparkCartesianSeries GetSeries()
        {
            switch (this.CallType)
            {
                case CallType.Duration:
                case CallType.HoldTime:
                    return new SparkLineSeries();
                case CallType.ResolvedIssues:
                    SparkWinLossSeries winLoss = new SparkWinLossSeries();
                    winLoss.ShowNegativePointIndicators = true;
                    return winLoss;
                default:
                    return new SparkLineSeries();
            }
        }

        private string GetValueMember(CallType callType)
        {
            switch (callType)
            {
                case CallType.Duration:
                    return "DurationSeconds";
                case CallType.HoldTime:
                    return "HoldTimeSeconds";
                case CallType.ResolvedIssues:
                    return "IssueResolved"; ;
                default:
                    return string.Empty;
            }
        }

        private void UpdateSeriesStyles()
        {
            if (this.CallType == CallType.ResolvedIssues)
            {
                this.alertLabel.Visible = false;
                this.timeLabel.Visible = false;
                this.radSparkline1.View.Margin = new Padding(0, 5, 0, 5);
            }
            else
            {
                this.alertLabel.Visible = true;
                this.timeLabel.Visible = true;
                this.radSparkline1.View.Margin = new Padding(5);
            }
        }

        protected override void Dispose(bool disposing)
        {
            if (disposing && (components != null))
            {
                components.Dispose();
            }

            this.timer.Tick -= this.Timer_Tick;

            base.Dispose(disposing);
        }
    }

    public enum CallType
    {
        Duration,
        HoldTime,
        ResolvedIssues
    }
}

File: Lib/Telerik/Examples/QuickStart/Sparkline/FirstLook/SparklineUserControl.vb
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Drawing
Imports System.Data
Imports System.Linq
Imports System.Text
Imports System.Windows.Forms
Imports Telerik.WinControls.UI
Imports Telerik.WinControls

Namespace Telerik.Examples.WinControls.Sparkline.FirstLook
    Partial Public Class SparklineUserControl
        Inherits UserControl

        Private _callType As CallType
        Private timer As Timer
        Private series As SparkCartesianSeries
        Private _owner As CalleeViewModel

        Public Sub New()
            InitializeComponent()
            Dim webUIFont As FontFamily = ThemeResolutionService.GetCustomFont("TelerikWebUI")
            Me.alertLabel.LabelElement.CustomFont = webUIFont.Name
            Me.alertLabel.LabelElement.CustomFontSize = 8
            Me.alertLabel.ForeColor = Color.Red
            Me.alertLabel.Text = ""
            Me.timeLabel.Font = New Font("Segoe UI", 9, FontStyle.Regular)
            Me.timer = New Timer()
            Me.timer.Interval = 750
            AddHandler Me.timer.Tick, AddressOf Me.Timer_Tick
        End Sub

        <Browsable(False)>
        <DesignerSerializationVisibility(DesignerSerializationVisibility.Hidden)>
        Public Property Owner As CalleeViewModel
            Get
                Return Me._owner
            End Get
            Set(ByVal value As CalleeViewModel)

                If Me._owner IsNot value Then
                    Me._owner = value
                End If
            End Set
        End Property

        <Browsable(False)>
        <DesignerSerializationVisibility(DesignerSerializationVisibility.Hidden)>
        Public Property CallType As CallType
            Get
                Return Me._callType
            End Get
            Set(ByVal value As CallType)

                If Me._callType <> value OrElse Me.radSparkline1.Series Is Nothing Then
                    Me.timer.[Stop]()
                    Me._callType = value
                    Me.CreateSeries()
                    Me.UpdateSeriesStyles()
                    Me.timer.Start()
                End If
            End Set
        End Property

        Public Overrides Function ToString() As String
            Select Case Me.callType
                Case callType.Duration
                    Return "Call Duration"
                Case callType.HoldTime
                    Return "Hold Time"
                Case callType.ResolvedIssues
                    Return "Issues Resolved"
                Case Else
                    Return MyBase.ToString()
            End Select
        End Function

        Protected Overrides Sub OnLoad(ByVal e As EventArgs)
            MyBase.OnLoad(e)
            Me.timer.Start()
        End Sub

        Private Sub Timer_Tick(ByVal sender As Object, ByVal e As EventArgs)
            If Me.owner Is Nothing Then
                Return
            End If

            Me.owner.AddRandomCall()

            If Me.callType = callType.Duration Then
                Me.alertLabel.LabelElement.Visibility = If(Me.owner.AverageDuration.Minutes >= 20, ElementVisibility.Visible, ElementVisibility.Hidden)
                Me.timeLabel.Text = Me.owner.AverageDuration.ToString()
                Me.timeLabel.ForeColor = If(Me.owner.AverageDuration.Minutes >= 20, Color.Red, Color.Empty)
            ElseIf Me.callType = callType.HoldTime Then
                Me.alertLabel.LabelElement.Visibility = If(Me.owner.AverageHoldTime.Minutes >= 10, ElementVisibility.Visible, ElementVisibility.Hidden)
                Me.timeLabel.Text = Me.owner.AverageHoldTime.ToString()
                Me.timeLabel.ForeColor = If(Me.owner.AverageHoldTime.Minutes >= 10, Color.Red, Color.Empty)
            End If
        End Sub

        Private Sub CreateSeries()
            Me.radSparkline1.Series = Nothing
            Me.series = Me.GetSeries()
            Me.series.ValueMember = Me.GetValueMember(Me.callType)
            Me.series.DataSource = Me.owner.Calls
            Me.radSparkline1.Series = series
        End Sub

        Private Function GetSeries() As SparkCartesianSeries
            Select Case Me.callType
                Case callType.Duration, callType.HoldTime
                    Return New SparkLineSeries()
                Case callType.ResolvedIssues
                    Dim winLoss As SparkWinLossSeries = New SparkWinLossSeries()
                    winLoss.ShowNegativePointIndicators = True
                    Return winLoss
                Case Else
                    Return New SparkLineSeries()
            End Select
        End Function

        Private Function GetValueMember(ByVal callType As CallType) As String
            Select Case callType
                Case CallType.Duration
                    Return "DurationSeconds"
                Case CallType.HoldTime
                    Return "HoldTimeSeconds"
                Case CallType.ResolvedIssues
                    Return "IssueResolved"
                Case Else
                    Return String.Empty
            End Select
        End Function

        Private Sub UpdateSeriesStyles()
            If Me.callType = callType.ResolvedIssues Then
                Me.alertLabel.Visible = False
                Me.timeLabel.Visible = False
                Me.radSparkline1.View.Margin = New Padding(0, 5, 0, 5)
            Else
                Me.alertLabel.Visible = True
                Me.timeLabel.Visible = True
                Me.radSparkline1.View.Margin = New Padding(5)
            End If
        End Sub

        Protected Overrides Sub Dispose(ByVal disposing As Boolean)
            If disposing AndAlso (components IsNot Nothing) Then
                components.Dispose()
            End If

            RemoveHandler Me.timer.Tick, AddressOf Me.Timer_Tick
            MyBase.Dispose(disposing)
        End Sub
    End Class

    Public Enum CallType
        Duration
        HoldTime
        ResolvedIssues
    End Enum
End Namespace
